### PR TITLE
Support custom endpoint for S3

### PR DESCRIPTION
**Pull request by mizeng**
Feature for https://github.com/grafana/loki/issues/699, support custom endpoint which is using `Path Style URL` of S3.

### CODE REVIEW

File: pkg/chunk/aws/s3_storage_client.go [hunk 1]
```diff
@@ -47,6 +47,10 @@ func NewS3ObjectClient(cfg StorageConfig, schemaCfg chunk.SchemaConfig) (chunk.O
 		return nil, err
 	}
 
+	if cfg.S3ForcePathStyle {
+		s3Config = s3Config.WithS3ForcePathStyle(true)
```
Comment (bboreham) [inline pkg/chunk/aws/s3_storage_client.go]:
```suggestion
		s3Config = s3Config.WithS3ForcePathStyle(cfg.S3ForcePathStyle)
```

(and remove the `if`)

File: pkg/chunk/aws/dynamodb_storage_client.go [hunk 2]
```diff
@@ -135,6 +136,9 @@ func (cfg *StorageConfig) RegisterFlags(f *flag.FlagSet) {
 
 	f.Var(&cfg.S3, "s3.url", "S3 endpoint URL with escaped Key and Secret encoded. "+
 		"If only region is specified as a host, proper endpoint will be deduced. Use inmemory:///<bucket-name> to use a mock in-memory implementation.")
+	f.BoolVar(&cfg.S3ForcePathStyle, "s3.force-path-style", false, "Set this to `true` to force the request to use path-style addressing "+
```
Comment (bboreham) [inline pkg/chunk/aws/dynamodb_storage_client.go]: this explanation is a bit long for here - maybe put it in `docs/arguments.md` ?

File: pkg/chunk/aws/s3_storage_client.go (hunk 1 above)
Comment (mizeng) [reply pkg/chunk/aws/s3_storage_client.go]: looks better. thanks

File: pkg/chunk/aws/dynamodb_storage_client.go (hunk 2 above)
Comment (mizeng) [reply pkg/chunk/aws/dynamodb_storage_client.go]: maybe a bit long here. But I don't think `docs/arguments.md` is a good place though. Some doc to describe the config.yaml maybe a good place, but I don't find one in Cortex.

Comment (mizeng) [reply pkg/chunk/aws/dynamodb_storage_client.go]: updated. plz help review again.

**Comment by mizeng** [comment]
could you help to merge this? I'm waiting for porting it back to Loki..

**Comment by bboreham** [comment]
Needs another approval per our [process](https://github.com/cortexproject/cortex/blob/master/GOVERNANCE.md#approving-prs)

**Pull request merged**